### ISSUE TITLE
Support an alternative to `$`-imports

### ISSUE THREAD

**Issue by lefou** [com-lihaoyi/mill#3062]
We inherited the various `$`-imports like `import $ivy` or `import $file` from Ammonite. It turns out, the Scala compiler isn't very fond of us using the `$` symbol in object and class names and newer versions starting from `3.4.0` enforce this more strictly. 

See also:
* https://github.com/lampepfl/dotty/issues/19702
* https://github.com/com-lihaoyi/Ammonite/pull/1394

I think a nice alternative to these magic imports could be something similar as the _using directives_  as known from Scala CLI.

From a compiler perspective, they are just comments, so we don't need to mangle scripts just for the sake of replacing directives with something regular. Something we needed to do before with the `$`-imports.

My preferred usage would look like this:

```scala
//> mill version 0.11.7
//> mill dep g:a:v
```

But since my proposal to make them tool/context specific wasn't accepted, the "standard" way to support tool specific key-value pair would be to encode the tool/context in the key. It would look like this:

```scala
//> using milll.version 0.11.7
//> using mill.dep g:a:v
```

The pros for following this standard would be, that IDEs, formatter and syntax highlighters might handle them better. (Which is already the case at the time of writing this. GitHub renders my first example just grey, but the second nicely colored.)

**Comment by lefou** [comment]
FTR, here is my mentioned proposal: https://github.com/scalacenter/tooling-working-group/discussions/68#discussioncomment-6945562

@lefou wrote on Sept 8, 2003:
> I've been playing around with using directive for a while, mostly in the context of Mill build scripts, and also had many discussions including those on the first Tooling Summit this year.
> 
> I suggest, to split the "using directives" (as known in it's current form) into **three** parts: the comment marker, the context marker and the directive, and discus each part, it's purpose and it's potential for standardization separately.
> 
>     1. **The comment marker `//>`**. This is undoubtedly agreed, that is nice. It's a comment, so it's transparent to the compiler. It's easy to recognize and survives most code formatting attempts. This is a nice candidate for a "standard".
> 
>     2. **The context marker**, currently the "using" keyword. There was lots of discussion about the "using" keyword, it's name and also the fact whether it's needed or not. I'd like to associate the context or the intepreting tools with it. Although a more generice "scala-cli" whould be more clear, "using" is now associated with Scala CLI and will do it. For other tools, there should be a different marker like "mill" or "ammonite"/"amm". That way, tool-specific configuration can co-exist and a script can be compatible to multiple tools/runners.
> 
>     3. **The context-specific directive or configuration**, typically some key-value like tool/context specific configuration. Due to it's tools specific nature, a more broader standard makes not much sense, and will make the evolution of the various tools harder.
> 
> 
> **Why do we need context specific configuration?**
> 
> Different tools have different features and properties. They also have different requirements, different defaults and different usability concepts. It's best to leave the control over the configuration of these details to the respective tools. Instead, let's make sure that tools can co-exists and scripts can be used in various environments simultaneously. With my proposal to introduce a context marker (or handle the "using" as marker for Scala-CLI only), it is possible to apply different configurations for different tools to the same script.
> 
> **Showcase**
> 
> Take the `build.sc` for example. It is an Scala script file and many different tools (Ammonite, Scala-CLI, Mill, xxx worksheet) can edit, format, execute it, but all will intepret it slightly different. Lets say I want to use the `os-lib` dependency, in Ammonite or Mill I can just use it, as it is already provided, whereas in Scala CLI I need to import it first (with `//> using dep`). If said script requires a minimal version of Ammonite or Scala-CLI to function, I could easily configure that due to the distinct context selectors (`//> using version` vs `//> ammonite version`).
> 
> Example: A script with configuration for Scala-CLI and Ammonite
> 
> ```
> //> using dep "com.lihaoy::os-lib:0.9.1"
> //> ammonite version 3.0.0
> 
> println("Current path: ${os.pwd}")
> ```
> 
> At a first glance, when opening this file, I can see it is supposed to run with Scala-CLI and/or Ammonite. It uses the os-lib dependency and requires Ammonite in version 3.0.0.
> 
> _See also my other post where I presented this idea: [com-lihaoyi/Ammonite#1372 (comment)](https://github.com/com-lihaoyi/Ammonite/discussions/1372#discussioncomment-6872317)_

**Comment by lefou** [comment]
> +1 for moving away from the $ivy imports. I tripped over this a few times with IntelliJ, where it decided to 'clean up' imports and messed up the $ivy imports.

A tip: Recently I have discovered, if you place the `$`-import after the other regular imports, IntelliJ seems to not remove them when organizing the imports. At least, in may case, it was constantly removing the `import $meta._` from a `build.sc` whenever a new import was about to be added, and I could prevent it by placing the `import $meta._` after all other imports.

**Comment by bishabosha** [comment]
I am also in favour of this - it would mean no custom parser needed for the scala 3 port (unless we would keep the old way specifically for deprecation messages)

**Comment by lihaoyi** [comment]
I'd like to preserve magic imports for backwards compatibility, at least for the first scala 3 release. It's fine to break stuff in a breaking release, but leaving the ivy imports in place for at least one or two releases should help smoothen the migration considerably

**Comment by lefou** [comment]
We could introduce using directives and deprecate imports in `0.12`, so that we could start clean in Mill `0.13`.

**Comment by lihaoyi** [comment]
Possible, but I would like to spend a bit more time exploring the design space for this area.

 directives work great to replace ivy imports, are kind of awkward to replace file imports. It's plausible we could find some solution, e.g. by making file imports go away by auto discovering files, but that may require other changes e.g. changing the file extension from .sc to .mill to avoid picking up random scripts. And I think there's some interesting stuff we could do with a more structured JSON/YAML header format, which could precisely match our JSON serialization format and generalize to arbitrary configs in a way that directives dont

That's not to say I think we should never do dirextives, but from a timeline perspective I think we should not couple it to the Scala 3 upgrade, as this area will take more time to work out the details

**Comment by lefou** [comment]
@lihaoyi Could you elaborate on the difficulties you see with `$file` imports? Since we already pre-preprocess them and replace with some dummy `import` statement. Their compiled binary representation is on the classpath and the actual `import` importing the package or class should work without any change.

**Comment by lefou** [comment]
Regarding the more complex encoding like JSON/YAML. I think this was discuss for using directives many times, finally it was dropped because of complexity. There is still no good editor support for such file headers. And as a result it's harder to properly edit, format and get the header right. In contrast, the current using directives are simple key value pairs and that's all. Everything else is parsing the value, which is analog to e.g. how CLI args work or sysprops or env values. 

I also envision to use a simple 

```scala
//> using mill.version 0.12.0
```

and want to parse this from `millw`, so that there is no longer a `.mill-version` file needed. As long as the format is that trivial, parsing in `bash`, `cmd` or PowerShell is easy.

**Comment by lefou** [comment]
@lihaoyi Do you have an opinion how we should parse the using directives? Should we use the same lib used by scala-cli, https://github.com/VirtusLab/using_directives, or rather parse it with fastparse, as we do for the `$`-imports.

**Comment by lihaoyi** [comment]
@lefou I'd like to hold off on committing to `using` directives for now. I think Mill's bootstrapped builds with JSON-serializable task values lends itself more to a different design using JSON/YAML/TOML which can directly correspond to meta-build task results. But it's not a major problem at the moment so would like to stick with `$ivy` imports until I have a chance to experiment with a more general design

**Comment by lefou** [comment]
My may focus is not replacing existing `$-imports` but supporting various other configs in the header of the `build.mill` instead of various cluttered dot-files. 

My candidates: 
* `.mill-version` -> `//> using mill.version`
* `.mill-jvm-version` -> `//> using mill.jvm.version`
* `.mill-opts`
* `.mill-jvm-opts`

Since we now have a native client, reading those should be fast enough.

**Comment by lihaoyi** [comment]
`mill-version` probably needs to stay in a separate file since it needs to be used from the bootstrap script written in Bash/Bat

The other files are indeed great candidates to move in. I still think a JSON/YAML/TOML frontmatter would be a better fit here. Depending on how we do this it shouldn't be a breaking change, so we don't need to be constrained by the 0.13.0 release timeline

**Comment by lefou** [comment]
The simple form of using directives make them rather easy to parse from a shell.

```sh
MILL_VERSION="$(head build.mill | grep '//> using mill.version ' | sed 's;//> using mill.version ;;')"
```

The batch script is probably a bit mor complicated.

**Comment by lihaoyi** [comment]
Yes there's definitely some room to play around, so feel free to explore. But let's hold off on committing to directive syntax for now

**Comment by lihaoyi** [comment]
More concretely, I think there are two other approaches that are worth exploring:

* Gradle puts its `plugins{}` block in-line using the same syntax as the main `build.gradle{.kts}`, that is parsed and separated out from the main build config before compilation. 
    * The upside is a consistent syntax for meta-build-level config and build-level config, 
    * The downside is the surprising evaluation model (e.g. it violates normal expectations around lexical scoping) and relative verbosity compared to magic imports or directives

* Python PEP723-style Inline Script Metadata. This is more similar to the directives idea proposed here and elsewhere, but uses TOML instead of Scala-CLI's bespoke directives syntax. 
    * One upside of TOML are its widespread familiarity even outside the Scala community (e.g. cargo.toml, pyproject.toml, etc.) 
    * Another upside is the fact that as a JSON-equivalent language, TOML it maps straightforwardly to Mill's serialized data structures. Thus any pure-data config (basic collections, primitives, and primitive-like things such as `ivy"..."` strings) can be trivially represented in TOML config without needing a separate mapping step (v.s. Scala-CLI directives which are more restricted and would need to be explicitly mapped to and from Mill task values)

**Comment by sideeffffect** [comment]
An alternative to TOML would be a (subset of) YAML (1.2). The advantage is that it's even more popular, more widespread and has wider support in tools etcthan TOML.

There are Scala libraries for working with YAML
 * https://github.com/VirtusLab/scala-yaml
 * https://github.com/circe/circe-yaml

**Comment by lefou** [comment]
I think we discuss two aspects here.

## General Mill configuration

This topic has it's own discussion (https://github.com/com-lihaoyi/mill/discussions/1226).  It's not really about embedding a full config into the `build.mill`. But some of the newer imports like `import $meta._` and `import $packages._` are nice, if they can be configured in-file. They are simple config switches and fall into that category. To replace their `$`-imports, we don't require any complex embedded config language. We should aim for something that is transparent to the scala compiler, so buildfiles work as other source files in any editor. The current `$-import`s don't work nicely in any other scala editor.

I don't think we need a config language that's able to hold complex values interoperable with tasks, since everything we can express as tasks, we should express as tasks. We only want some small values typically configured via CLI options (e.g. `--jobs`) or environment variables, for which simply key-value pairs are enough.

## Configuration of the meta-build without creating or touching the meta-build

That's what all the other `$-import` statements are about. It's mostly adding plugins or files to the classpath. Since we already have a simple representation for ivy-dependencies, that works as string interpolation (`ivy"g:a:v"`) as well as on the CLI (`--import ivy:g:a:v`), we don't need some complex config language either. 

As we don't seek some general purpose configuration scheme to be embedded into a `build.mill` but only require some simple key-value pairs, we should find a mechanism that doesn't add inconvenience to the `build.mill` files. (E.g. I'd like to open them with any Scala editor at hand (like `kate` or `nano`) with working syntax highlighting.) Anything more complex than a single-line comment will break this premise, hence I think `using directives` as specified by the scala-cli are "good enough".

**Comment by lihaoyi** [comment]
@sideeffffect yes YAML is definitely worth considering. It's basically the same as the PEP 723 TOML proposal with a slightly different language, but is also widely used e.g. as YAML frontmatter in documentation systems.

Regarding Scala-CLI using directives, I do not like them and said as much in the original discussions (https://contributors.scala-lang.org/t/pre-sip-using-directives/5700, https://contributors.scala-lang.org/t/pre-sip-scala-cli-as-new-scala-command/5628). If we just need a key-value config syntax, I want one with broader support and usage in the wider programming ecosystem. https://github.com/VirtusLab/using_directives does not fit that bill

**Comment by lefou** [comment]
> I want one with broader support and usage in the wider programming ecosystem.

@lihaoyi you stated before, that you don't think we need to innovate. So we need to use something existing, but there is not much to choose from. Any config-, preamble- or other kind of pre-processor stuff that is not compatible with Scala syntax is out of scope. So we need to concentrate on stuff for languages that have same comment syntax as Scala, and it doesn't look like any of these languages do have a concept to configure some runtime stuff in the source files. (There is `doxygen` to generate docs from sources, which uses some Scaladoc-like comments including `//>` with `@`-tags.) 

I don't fully like the using directives either. I even tried to change some bits of it while in the design phase, before it went final (e.g. in the Scala Tooling summit March 2023), but didn't succeed. 

As for your issues with using directives you linked to. Let me note, we don't speak about designing the Scala language here, it's only about some convenience for Mill, a rather niche build tool.

> 1. It’s re-purposing an existing keyword in a weird way. using is kinda-sorta tangentially related to one particular use case of what we’re trying to do here, just like import $ is kinda-sorta tangentially related to one particular use case of what Ammonite does. But it’s different enough that it seems like it would cause confusion, not unlike how we used to use underscore _ for a bunch of scenarios that were kinda-sorta related

Since then, all using directives are comments, and don't loo like syntax anymore. It's also a standard now. It already exists and users need to deal with it, disregarding whether we use it in Mill or not. 

> 2. It doesn’t look like Scala. scala "3.1.2" does not look like valid Scala syntax at all. someSettings { setting1 value 1; setting2; } is technically valid Scala syntax, but is in the style of odd DSLs. It looks like Kotlin or Groovy with their convention of builder DSLs, but those aren’t ubiquitous in the Scala ecosystem.

This part never made it into the final version. There are no `{}`-blocks and everything is comments now.

> 3. The fact that it looks similar to Scala expressions at all is misleading. Can I import stuff? What’s the evaluation order? Can I define vals/defs/classes and so on? I assume not, and that this stuff would get statically inspected as metadata, rather than evaluated as code. This means the syntax is actively misleading people into thinking it’s Scala expressions, when it’s not

In the final version, using directives are no Scala-like syntax at all but just a special comment with the additional "using" match-word. So, there is no confusion left. All we have is a comment with a "using" in it and some key-value pair, which is pretty easy to read and also easy to search for in the internet.


I think the only other viable way to get some meta-config into the sources in respect to the Scala language would be __annotations__. But, there is no annotation support for files, so we can't have the annotation at the top of the file. I used that way in `sbuild`, another Scala build tools attempt. It has a lot of downsides. The biggest is, you probably need a full Scala parser and also some runtime machinery to verify all you other constraints, to makes sure to detect duplicates and contradicting settings. It's not easy at all to extract those values without starting up a real tool, e.g. a Mill config parser or Mill itself. It's nothing to be done easily from a `bash` or `cmd` script. And nothing easily added to various other tools. 

Again, I'm not speaking about dedicated config format we want to use for a `mill.conf` file here.

Any other structured config format like YAML, TOML, JSON (maybe with reduced feature sets) are out of scope to be embedded in a Scala source file. Simply, because there is no ubiquitous standard to prepend a source-file with some meta-data that is supported by any editor and formatter on all platforms.

**Comment by lefou** [comment]
* `import $ivy` is now replaced by the YAML frontmatter `//| mvnDeps:` attribute.
* `import $file`was silently dropped with Mill 0.12, I think
* `import $packages._` and `$meta._` are gone completely, Mill now enables these feature by default.
